### ISSUE TITLE
LFX Proposal: Upgrade the documentation publishing pipeline for Kubescape controls

### ISSUE THREAD

**Issue by craigbox** [kubescape/kubescape#1302]
This issue relates to a **proposal** for a [LFX mentorship project](https://github.com/cncf/mentoring/tree/main/programs/lfx-mentorship/2023/03-Sep-Nov) for Term 3, 2023.  **It has not yet been accepted**. Please feel free to comment on this issue about how we might improve the proposal, but **commenting here is not the same as applying to LFX**.

- Description: [Kubescape's control library](https://github.com/kubescape/regolibrary) includes more than 200 controls, tests that codify Kubernetes best practices derived from the most prevalent security frameworks in the industry. Metadata in the controls is used to generate documentation pages in the ARMO website. This project will update this automation to make this control documentation available on kubescape.io.
- Expected Outcome: A full set of documentation for Kubescape controls on kubescape.io. Stretch goals include better README-style documentation inside the repository, and documentation pages on how the controls, frameworks and tests relate.
- Recommended Skills: 
  - Python
  - Technical writing
  - Rego
- Mentors:
  - Craig Box (@craigbox, craigb AT armosec.io)
  - Ben Hirschberg (@slashben, ben AT armosec.io)

**Comment by PurviChaurasia** [comment]
Hi @craigbox! I am extremely interested in this project and would like to get a head start into the application process. I have previously worked with python and have also done documentation for various in past internships. Could you perhaps guide me on how to make my application stronger?

**Comment by craigbox** [comment]
We can't guide individuals on how to write their applications, but we can say what we look for. I would hope to find a candidate who can demonstrate excellent technical writing skills alongside proof they will be able to do a good job programming.  

In the last two terms I've had three buckets of candidates: people who have submitted a generic cover letter which could apply to anything, people who at least seem to understand what Kubescape is and what the proposal asks, and people who have actually thought about the problem and documented some possible lines of enquiry, or how they might go about their project.  We've taken candidates exclusively from the last bucket.

**Comment by PurviChaurasia** [comment]
Alright, thanks a lot! Will aim to be in the last bucket of candidates!

**Comment by PurviChaurasia** [comment]
Hi @craigbox, just wanted to clarify a few things regarding the application. Apart from the cover letter and resume, what else does kubescape require as part of the application process? Do we need to work on a proposal for the project as well?

**Comment by craigbox** [comment]
We don't require anything, but all these things help make you stand out as a candidate.

**Comment by salty-ivy** [comment]
@craigbox, I'm very excited about this project! I have a good amount of experience with Python mostly as contributor in @wagtail, @django and @plone, and I've worked on documentation projects before. Just to confirm, the main goal of this project is to update the automation to make the Kubescape documentation available on kubescape.io, right?

I have a suggestion for the implementation. How about using `sphinx` for generating the documentation? We could create a separate repository specifically for the documentation, which would serve as the source of updates. We can set up a `github workflow` to automate the process, and additionally, use a `makefile` to allow for easy manual execution of the documentation generation using the `make` command.

This way, we can ensure that the documentation stays up-to-date and is easily accessible on kubescape.io. I'm eager to hear your thoughts on this approach!

**Comment by goyalpramod** [comment]
@salty-ivy fantastic idea, I had a similar approach in my mind. But do you not believe working with `mkdocs`  will be better and also more accessible from collaboration point of view?

**Comment by salty-ivy** [comment]
yea

> @salty-ivy fantastic idea, I had a similar approach in my mind. But do you not believe working with `mkdocs` will be better and also more accessible from collaboration point of view?

yes I think so !

**Comment by craigbox** [comment]
Changing kubescape.io away from `mkdocs` is not in scope.  Sections that need to be generated with a different tool (for example, [Swagger](https://swagger.io/tools/swagger-ui/), can either be [integrated](https://blueswen.github.io/mkdocs-swagger-ui-tag/pet-store/) or hosted in parallel (think api.kubescape.io)

**Comment by PurviChaurasia** [comment]
I agree, shifting to a completely different software for this purpose will not be very suitable. For specific use cases can be done so using the Swagger API integrated with `mkdocs` as mentioned by @craigbox. A more simpler approach I believe would be to setup a `Github Workflow` under `kubescape.io`, set to run every time push is made to the `rego library`.

**Comment by PurviChaurasia** [comment]
I believe the part we would have to work on, is that since `mkdocs` functions only via markdowns, we'll have to convert the metadata of the controls library to a markdown format.

**Comment by PurviChaurasia** [comment]
A rather basic question that perhaps all of you would be able to help me understand - We will be working on the generation of the [controls](https://github.com/kubescape/regolibrary/tree/master/controls) library via its json files right?
(I might be completely wrong hence wanted confirmation)

**Comment by goyalpramod** [comment]
@PurviChaurasia I believe that is the case, it seems. The controls library is currently codified in JSON files, and metadata within these controls is used to generate documentation pages on the ARMO website. The project's goal is to enhance this process so that the control, tests documentation is accessible on kubescape.io. This likely involves improving the automation that extracts information from the JSON files and presents it in a user-friendly format on the website.

**Comment by PurviChaurasia** [comment]
Yes that is what I wanted to confirm, thanks a lot @goyalpramod !

**Comment by PurviChaurasia** [comment]
So perhaps, then our main task becomes extracting and converting the `json` files into `markdown` format as that is the one supported by `mkdocs`. From what I can gather, there are various ways of doing that ranging from built-in plugins with `mkdocs` or even a simple `python` script. This can then be added to the `Github Workflow`.

**Comment by dev-hack95** [comment]
Hey Ben are in charge of kubespace admission pod controller creation project right
I have a idea to create a admission controller
We will create mutating webhook for the controller for that we need a cert auth we can use any third party cert auth anfd then we will create a tls yaml file to make sure that controller will recognize the kubespcae controller
After that we can create a webhook configuration  in yaml file with ca Certifacate so that it will know how to validate tls certificate and we can add some rules to trigger the request to webhook
We will write code to auth using kube config and we use client varibale or set to talk to controller
We will create mutating end points and deploy to k8s\

**Comment by XDRAGON2002** [comment]
@dev-hack95 Just a gentle request to kindly not spam the thread with the same comments, as in most situations it just sends out random notifs to all the people in the thread.

Both on this issue and the other one, though to redirect you, this issue is completely separate from the one you're referring to, kindly take a look at #1301 for that.

**Comment by craigbox** [comment]
Thanks everyone; we're pleased to announce that Rohit ([TRohit20](https://github.com/TRohit20)) has been selected to take on this project!
